wenet/efficient_conformer/encoder.py

```diff
@@ -381,32 +396,42 @@ def forward_chunk(
         else:
             next_cache_start = max(attention_key_size - required_cache_size, 0)
 
-        # for ONNX export， padding xs to chunk_size
-        if self.global_chunk_size > 0:
-            real_len = xs.size(1)
-            xs = F.pad(xs, (0, 0, 0, self.global_chunk_size - real_len), value=0.0)
-            tmp_zeros = torch.zeros(att_mask.shape, dtype=torch.bool)
-            att_mask[:, :, required_cache_size + real_len + 1:] = \
-                tmp_zeros[:, :, required_cache_size + real_len + 1:]
-
         r_att_cache = []
         r_cnn_cache = []
         mask_pad = torch.ones(1,
                               xs.size(1),
                               device=xs.device,
                               dtype=torch.bool)
         mask_pad = mask_pad.unsqueeze(1)    # batchPad (b=1, 1, time=chunk_size)
+
+        if self.global_chunk_size > 0:
+            # for ONNX decode simulation
+            pos_emb = self.embed.position_encoding(
+                offset=max(offset - cache_t1, 0),
+                size=cache_t1 + self.global_chunk_size)
+            att_mask[:, :, -self.global_chunk_size:] = chunk_masks
+            mask_pad = chunk_masks.to(torch.bool)
+        else:
+            pos_emb = self.embed.position_encoding(
+                offset=offset - cache_t1, size=attention_key_size)
+
         max_att_len, max_cnn_len = 0, 0     # for repeat_interleave of new_att_cache
         for i, layer in enumerate(self.encoders):
             factor = self.calculate_downsampling_factor(i)
             # NOTE(xcsong): Before layer.forward
             #   shape(att_cache[i:i + 1]) is (1, head, cache_t1, d_k * 2),
             #   shape(cnn_cache[i])       is (b=1, hidden-dim, cache_t2)
             # shape(new_att_cache) = [ batch, head, time2, outdim//head * 2 ]
+            att_cache_trunc = 0
+            if xs.size(1) + att_cache.size(2) / factor > pos_emb.size(1):
+                # The time step is not divisible by the downsampling multiple
+                # We propose to double the chunk_size.
+                att_cache_trunc = xs.size(1) + \
+                    att_cache.size(2) // factor - pos_emb.size(1) + 1
             xs, _, new_att_cache, new_cnn_cache = layer(
                 xs, att_mask, pos_emb,
                 mask_pad=mask_pad,
-                att_cache=att_cache[i:i + 1, :, ::factor, :],
+                att_cache=att_cache[i:i + 1, :, ::factor, :][:, :, att_cache_trunc:, :],
                 cnn_cache=cnn_cache[i, :, :, :]
                 if cnn_cache.size(0) > 0 else cnn_cache
             )
```

Comment:
Q1: 

```
xs.size(1) + att_cache.size(2) / factor > pos_emb.size(1)
```
OR
```
( xs.size(1) + att_cache.size(2) ) / factor > pos_emb.size(1)
```
?
results of the above two lines are not equal :

![image](https://user-images.githubusercontent.com/13466943/220524358-e616afe6-279e-46ae-b5d1-9cb646dbc880.png)


Q2:

What do mean by `double the chunk_size` ? I think `[:, :, att_cache_trunc:, :]` will simply eliminate any unnecessary attention cache information at the beginning, so where is the `double` ?

Reply:
Q1: `xs.size(1) + att_cache.size(2) / factor > pos_emb.size(1)`, because `xs` was downsampled in the previous block.

Q2: For the `train_u2++_efficonformer_v2.yaml`, downsample rate is 1/2 (conv2d2) + 1/4 (efficonformer block), so 18->36 can reduce the downsampling loss.

The description of double is really inaccurate here, so let me adjust it.